test/langtools/tools/javac/modules/AnnotationFilerTest.java

```diff
@@ -0,0 +1,191 @@
+            """
+            package lib;
+            public class LibClass {
+                public static final String FIELD = "";
+            }
+            """;
+
+    // Target source references the field to force library compilation.
+    private static final String TARGET_SOURCE =
+            """
+            class TargetClass {
+                static final String VALUE = lib.LibClass.FIELD;
+            }
+            """;
+
+    private static final String LIB_CLASS_TYPE_NAME = "lib.LibClass";
+    private static final Path LIB_JAR = Path.of("lib.jar");
+
+    private static final String LIB_SOURCE_FILE_NAME = "lib/LibClass.java";
+    private static final String LIB_CLASS_FILE_NAME = "lib/LibClass.class";
+
+    // These differ only in the filename suffix part and represent the trailing
+    // parts of a Path URI for source/class files in the JAR. The source file
+    // will exist, but the class file must not be created by annotation processing.
+    private static final String JAR_SOURCE_URI_SUFFIX = "lib.jar!/" + LIB_SOURCE_FILE_NAME;
+    private static final String JAR_CLASS_URI_SUFFIX = "lib.jar!/" + LIB_CLASS_FILE_NAME;
+
+    @Test
+    public void classpathJarsCannotBeWrittenDuringProcessing() throws IOException {
+        ToolBox tb = new ToolBox();
+        tb.createDirectories("lib");
+        tb.writeFile(LIB_SOURCE_FILE_NAME, LIB_SOURCE);
+        new JarTask(tb, LIB_JAR).files(LIB_SOURCE_FILE_NAME).run();
+
+        // These are assertions about the test environment, not the code-under-test.
+        try (FileSystem zipFs = FileSystems.newFileSystem(LIB_JAR)) {
+            // The bug would have only manifested with writable JAR files, so assert that here.
+            assertFalse(zipFs.isReadOnly());
+            // This is the JAR file URI for the *source* code. Later we attempt to create the
+            // sibling *class* file from within the annotation processor (which MUST FAIL).
+            // We get the source URI here to verify the naming convention of the JAR file
+            // URIs to prevent the _negative_ test we do later being fragile.
+            URI libUri = zipFs.getPath(LIB_SOURCE_FILE_NAME).toUri();
+            assertEquals("jar", libUri.getScheme());
+            assertTrue(libUri.getSchemeSpecificPart().endsWith(JAR_SOURCE_URI_SUFFIX));
+        }
+
+        // Code under test:
+        // Compile the target class with library source only available in the JAR. This should
+        // succeed, but MUST NOT be able to create files in the JAR during annotation processing.
+        new JavacTask(tb)
+                .processors(new TestAnnotationProcessor())
+                .options("-implicit:none", "-g:source,lines,vars")
+                .sources(TARGET_SOURCE)
+                .classpath(LIB_JAR)
+                .run()
+                .writeAll();
+    }
+
+    @SupportedAnnotationTypes("*")
+    static class TestAnnotationProcessor extends AbstractProcessor {
```

Comment:
Please see the test library file

langtools/toosl/javac/lib/JavacTestingAbstractProcessor.java

and consider extending that class instead.